src/common/threading.jl

```diff
@@ -1,42 +1,92 @@
 import FFTW
 using LinearAlgebra
 
-function setup_threading(; n_fft=1, n_blas=Threads.nthreads())
-    n_julia = Threads.nthreads()
+function setup_threading(; n_fft=1, n_blas=Threads.nthreads(), n_DFTK=nothing)
+    if n_DFTK != nothing
+        set_DFTK_threads!(n_DFTK)
+    end
+    n_DFTK = @load_preference("DFTK_threads", Threads.nthreads())
     FFTW.set_num_threads(n_fft)
     BLAS.set_num_threads(n_blas)
-    mpi_master() && @info "Threading setup:" n_fft n_blas n_julia
+    mpi_master() && @info "Threading setup: $(Threads.nthreads()) Julia threads, $n_DFTK DFTK threads, $n_fft FFT threads, $n_blas BLAS threads"
+
 end
 
 """
 Convenience function to disable all threading in DFTK and assert that Julia threading
 is off as well.
 """
 function disable_threading()
-    n_julia = Threads.nthreads()
-    n_julia > 1 && mpi_master() && error(
-        "Julia currently uses $n_julia threads. Ensure that the environment variable " *
-        "JULIA_NUM_THREADS is unset and julia does not get the `-t` flag passed."
-    )
-    @assert n_julia == 1  # To exit in non-master MPI nodes
-    setup_threading(;n_fft=1, n_blas=1)
+    setup_threading(;n_fft=1, n_blas=1, n_DFTK=1)
 end
 
-# Parallelization loop breaking range into chunks.
-function parallel_loop_over_range(fun, storages::AbstractVector, range)
-    chunk_length = cld(length(range), Threads.nthreads())
+function set_DFTK_threads!(n)
+    if n > Threads.nthreads()
+        error("You're asking for $n_DFTK DFTK threads, but only ran julia with $(Threads.nthreads()) threads.")
+    end
+    if @load_preference("DFTK_threads", nothing) != n
+        @info "DFTK_threads preference changed. This is a permanent change, restart julia to see the effect."
+    end
+    @set_preferences!("DFTK_threads" => n)
+end
+function set_DFTK_threads!()
+    @delete_preferences!("DFTK_threads")
+end
+# If unset, use nthreads() threads
+function get_DFTK_threads()
+    nthreads = @load_preference("DFTK_threads", nothing)
+    if isnothing(nthreads)
+        nthreads = Threads.nthreads()
+    elseif nthreads > Threads.nthreads()
+        error("You set your preference for $n_DFTK DFTK threads, but only ran julia with $(Threads.nthreads()) threads.")
```

Comment:
Does such a crash also happen in case the `get_DFTK_threads()` function is not called ? Calling that function might not always be the case.